### ISSUE TITLE
Hashseed and cache miss for same queryset

### ISSUE THREAD

**Issue by ali-faizan** [Suor/django-cacheops#260]
I am having lots of cache misses. After debugging i saw that sql queries generated by same queryset have different order in where clause. This issue has been reported on Django:
[Same queryset result in two different queries on ORM](https://code.djangoproject.com/ticket/28297)

Setting PYTHONHASHSEED=2 does seem to resolve issue as mentioned in the ticket but i can't figure out how these 2 are related. Neither I know when the Django issue will get fixed as its been open for 5 months. Is there anyway we can fix this for Cacheops? As of now only thing i can think of is move to python 3.6 where dictionary order remain preserved.

Stack:
Python  3.5.4
Django 1.8.18
Cacheops 4.0.2

**Comment by Suor** [comment]
Are you sure the link you provided is related. It seems there query is broken completely sometimes not just some order of clause's changes.

Please provide your own example.

**Comment by ali-faizan** [comment]
I have this very simple model
```
class CSV_Import_Integration(Model):
    school_id = models.IntegerField()
    is_active = models.BooleanField(default=False)
    csv_path = models.CharField(max_length=255)
    dump_hash = models.CharField(max_length=255)
    date_updated = models.DateTimeField(auto_now=True)
    import_type = models.IntegerField()
```

And this is the code
```
models.CSV_Import_Integration.objects.filter(
            school_id=self.context.school_id, is_active=True,
            import_type=import_type.value
        ).order_by('-date_updated').first()
```

I executed above code 5 times and it generated following SQL. See the order difference in WHERE clause.

SELECT `webservice_csv_import_integration`.`id`, `webservice_csv_import_integration`.`school_id`, `webservice_csv_import_integration`.`is_active`, `webservice_csv_import_integration`.`csv_path`, `webservice_csv_import_integration`.`dump_hash`, `webservice_csv_import_integration`.`date_updated`, `webservice_csv_import_integration`.`import_type` FROM `webservice_csv_import_integration` WHERE (`webservice_csv_import_integration`.`school_id` = 12358 AND `webservice_csv_import_integration`.`is_active` = 1 AND `webservice_csv_import_integration`.`import_type` = 0) ORDER BY `webservice_csv_import_integration`.`date_updated` DESC LIMIT 1; args=(12358, True, 0)

SELECT `webservice_csv_import_integration`.`id`, `webservice_csv_import_integration`.`school_id`, `webservice_csv_import_integration`.`is_active`, `webservice_csv_import_integration`.`csv_path`, `webservice_csv_import_integration`.`dump_hash`, `webservice_csv_import_integration`.`date_updated`, `webservice_csv_import_integration`.`import_type` FROM `webservice_csv_import_integration` WHERE (`webservice_csv_import_integration`.`is_active` = 1 AND `webservice_csv_import_integration`.`school_id` = 12358 AND `webservice_csv_import_integration`.`import_type` = 0) ORDER BY `webservice_csv_import_integration`.`date_updated` DESC LIMIT 1; args=(True, 12358, 0)

SELECT `webservice_csv_import_integration`.`id`, `webservice_csv_import_integration`.`school_id`, `webservice_csv_import_integration`.`is_active`, `webservice_csv_import_integration`.`csv_path`, `webservice_csv_import_integration`.`dump_hash`, `webservice_csv_import_integration`.`date_updated`, `webservice_csv_import_integration`.`import_type` FROM `webservice_csv_import_integration` WHERE (`webservice_csv_import_integration`.`is_active` = 1 AND `webservice_csv_import_integration`.`import_type` = 0 AND `webservice_csv_import_integration`.`school_id` = 12358) ORDER BY `webservice_csv_import_integration`.`date_updated` DESC LIMIT 1; args=(True, 0, 12358)

SELECT `webservice_csv_import_integration`.`id`, `webservice_csv_import_integration`.`school_id`, `webservice_csv_import_integration`.`is_active`, `webservice_csv_import_integration`.`csv_path`, `webservice_csv_import_integration`.`dump_hash`, `webservice_csv_import_integration`.`date_updated`, `webservice_csv_import_integration`.`import_type` FROM `webservice_csv_import_integration` WHERE (`webservice_csv_import_integration`.`import_type` = 0 AND `webservice_csv_import_integration`.`school_id` = 12358 AND `webservice_csv_import_integration`.`is_active` = 1) ORDER BY `webservice_csv_import_integration`.`date_updated` DESC LIMIT 1; args=(0, 12358, True)

SELECT `webservice_csv_import_integration`.`id`, `webservice_csv_import_integration`.`school_id`, `webservice_csv_import_integration`.`is_active`, `webservice_csv_import_integration`.`csv_path`, `webservice_csv_import_integration`.`dump_hash`, `webservice_csv_import_integration`.`date_updated`, `webservice_csv_import_integration`.`import_type` FROM `webservice_csv_import_integration` WHERE (`webservice_csv_import_integration`.`school_id` = 12358 AND `webservice_csv_import_integration`.`import_type` = 0 AND `webservice_csv_import_integration`.`is_active` = 1) ORDER BY `webservice_csv_import_integration`.`date_updated` DESC LIMIT 1; args=(12358, 0, True)

**Comment by Suor** [comment]
Looks like Pythons from 3.3 to 3.5 have this issue. Looks like you can fix that by patching `Q` constructor:
```python
import sys
from django.db.models.query_utils import Q

if (3, 3) <= sys.version_info < (3, 6):
    def Q__init__(self, *args, **kwargs):
        super(Q, self).__init__(children=list(args) + list(sorted(kwargs.items())))
    Q.__init__ = Q__init__
``` 

Can you check if that works for you? If it will I will probably add it to cacheops until Django supports python 3.5.

**Comment by ali-faizan** [comment]
it worked! Thanks alot @Suor

**Comment by Suor** [comment]
Fixed in 5040816
